[SHK-60] Feed 등록, 수정, 삭제 기능 구현

Kim-Changgyu opened the pull request:
### ⛏ 작업 사항
- Feed, FeedTag 엔티티 구현
- 각 레이어별 등록/수정/삭제 관련 메서드 구현
- 태그 추출 클래스 구현 (등록/수정시 사용)
- 피드 이미지 등록 연동 구현

### 📝 작업 요약
- 피드 등록, 수정, 삭제 기능 구현

### 💡 관련 이슈
- Resolved : [SHK-109], [SHK-110], [SHK-111], [SHK-112], [SHK-113], [SHK-114]

[SHK-109]: https://shoekream.atlassian.net/browse/SHK-109?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-110]: https://shoekream.atlassian.net/browse/SHK-110?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-111]: https://shoekream.atlassian.net/browse/SHK-111?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-112]: https://shoekream.atlassian.net/browse/SHK-112?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-113]: https://shoekream.atlassian.net/browse/SHK-113?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-114]: https://shoekream.atlassian.net/browse/SHK-114?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Kim-Changgyu committed:
Feat : [SHK-109] 피드 등록 기능 개발

- StyleFacade에서 사용하는 MemberService의 메소드명 동기화 필요

Kim-Changgyu committed:
Feat : [SHK-110] 피드 삭제 기능 개발

- FK 옵션이 없기 때문에 delete에서 관련 테이블의 데이터를 일일이 삭제하도록 구현

Kim-Changgyu committed:
Feat : [SHK-111] 피드 수정 기능 개발

- 현재 개발된 기능 중에 수정 가능한 필드는 "본문" (추후 추가될 수 있음)

Kim-Changgyu committed:
Feat : [SHK-112] 피드 등록 및 수정시 태그(#) 처리 기능 추가

- 태그 관련 Insert & Delete 쿼리가 N개 발생하여 개선 필요

Kim-Changgyu committed:
Feat : [SHK-113] 피드 등록시 이미지 업로드 기능 추가

- 이미지 저장 기능 연동 [SHK-68]
- Multipart/form-data 형식으로 받기 때문에 value가 객체일 경우 잠재적으로 문제될 가능성이 높아보임

Kim-Changgyu committed:
Test : [SHK-114] 피드 관련 단위 및 통합 테스트

- 삭제 기능은 필요한 시점에 추가 예정

ttaehee commented on src/main/java/com/prgrms/kream/domain/style/controller/StyleController.java:
```diff
@@ -1,4 +1,52 @@
+		);
+	}
+
+	@DeleteMapping("/{id}")
+	public ApiResponse<String> delete(@PathVariable long id) {
+		styleFacade.delete(id);
+		return ApiResponse.of(
+				"성공적으로 삭제 됐습니다."
```
문자열 대신 ResponseStatus만 보내도 되지 않을까요?
그리고 ResponseStatus도 register()와 통일성을 위해 명시해주는게 좋을것 같아요!
위에 update()도요!

ttaehee commented on src/main/java/com/prgrms/kream/domain/style/dto/UpdateFeedRequestOfService.java:
```diff
@@ -0,0 +1,12 @@
+public record UpdateFeedRequestOfService(
+		String content
+) {
+	@Builder
+	public UpdateFeedRequestOfService(String content) {
+		this.content = content;
+	}
```
record 클래스인데 builder로 생성자를 만든 이유가 있을까요?!

Kim-Changgyu replied:
레이어에서 직접 요청/응답을 생성해서 상, 하위 레이어로 뿌릴 때 포맷팅하기 좋을 것이라고 판단해서 넣었으나
해당 요청/응답을 Mapper의 메서드로 빼면서 굳이 필요할 것 같지 않아 뺐습니다.

(아직 해당 부분에 대해 빌더를 사용하자는 룰이 정해지지 않은 이유도 있습니다 ㅎㅎ)

ttaehee commented on src/main/java/com/prgrms/kream/domain/style/facade/StyleFacade.java:
```diff
@@ -1,4 +1,54 @@
+	@Transactional
+	public FeedResponse register(CreateFeedRequestOfFacade request) {
+		FeedResponse response = styleService.register(
+				CreateFeedRequestOfService.builder()
+						.content(request.content())
+						.author(memberService.getMember(request.author()))
+						.build());
```
mapper 클래스가 이미 있으니 요 코드는 mapper의 메서드로 작성해도 좋을듯 합니다!

ttaehee reviewed (approved):
고생하셨습니다 😆!

BnDC commented on src/main/java/com/prgrms/kream/domain/style/dto/CreateFeedRequestOfFacade.java:
```diff
@@ -0,0 +1,7 @@
+
+public record CreateFeedRequestOfFacade(
+		String content,
+		Long author
```
`@NotBlank` 같은 Bean validation을 사용하는 것이 어떨까요?

Kim-Changgyu replied:
헉;; Validation 빠졌네요 이런 애매추어 보이 같으니라구...
바로 반영하겠습니다 감사합니다

BnDC commented on src/main/java/com/prgrms/kream/domain/style/service/StyleService.java:
```diff
@@ -23,4 +23,14 @@ public FeedResponse register(CreateFeedRequestOfService request) {
 
 		return StyleMapper.toDto(savedFeed);
 	}
+
+	@Transactional
+	public void delete(long id) {
+		feedRepository.findById(id)
+				.ifPresent(feed -> {
+					// 관련 테이블의 레코드 삭제 (Cascade)
+					feedRepository.delete(feed);
+				});
+	}
```
다른 부분은 id에 해당하는 데이터를 찾지 못했을 때 예외를 던져주는데 이 부분은 던지지 않는 이유가 있을까요?

Kim-Changgyu replied:
없는 데이터를 삭제하려고 하면 무시되게끔 구현했습니다 :)

BnDC reviewed (commented):
구현하느라 고생하셨습니다 몇 가지 코멘트 남겼습니다

BnDC commented on src/main/java/com/prgrms/kream/domain/style/controller/StyleController.java:
```diff
@@ -1,4 +1,52 @@
+
+	@PostMapping
+	@ResponseStatus(code = HttpStatus.CREATED)
+	public ApiResponse<FeedResponse> register(@ModelAttribute CreateFeedRequestOfFacade request) {
+		return ApiResponse.of(
+				styleFacade.register(request)
+		);
```
`@ModelAttribute` 로 받는 이유가 있을까요?

Kim-Changgyu replied:
프론트에서 이미지가 포함된 등록 정보를 보낼 경우 Content-Type이 Multipart/form-data 넘어옵니다 !
key-value 형식으로 넘어오는데 각 key에 해당하는 value를 @RequestPart로 받을 수 있지만 컨트롤러에서 너무 많은 인자가 노출되기 때문에 @ModelAttribute로 받도록 했습니다.

Kim-Changgyu committed:
Refactor : 피드백 반영 및 컨벤션 적용

피드백
- DTO Builder 제거
- Mapper 클래스 메서드 추가
- Validation 추가

컨벤션
- 응답 코드 명시적으로 표기
- DTO 레이어별 명칭 변경

BnDC reviewed (approved):
고생하셨어요~!
